### ISSUE TITLE
Fix S2583 FP: Reachable paths in foreach loops

### ISSUE THREAD

**Issue by E4est** [SonarSource/sonar-dotnet#8376]
### Description

[S2583](https://sonarsource.github.io/rspec/#/rspec/S2583/csharp) is raised on `if`-conditions inside of loops although one of the operands is manipulated inside of the loop.

### Repro steps

```csharp
const int NullsBeforeFeedback = 50;

int numberOfNulls = 0;

foreach (object? test in Stuff.Items)
{
    if (test != null)
    {
        continue;
    }

    numberOfNulls++;

    // Raised here.
    if (numberOfNulls % NullsBeforeFeedback == 0)
    {
        Console.WriteLine("placeholder");
    }
}

numberOfNulls = 0;

foreach (object? test in Stuff.Items)
{
    // It doesn't matter how it's structured.
    if (test == null)
    {
        numberOfNulls++;

        // Raised here.
        if (numberOfNulls % NullsBeforeFeedback == 0)
        {
            Console.WriteLine("placeholder");
        }
    }
}

numberOfNulls = 0;

foreach (object? test in Stuff.Items)
{
    // It doesn't matter how it's structured.
    if (test == null)
    {
        numberOfNulls++;

        // Raised here.
        if (numberOfNulls % NullsBeforeFeedback != 0)
        {
            continue;
        }

        Console.WriteLine("placeholder");
    }
}

static class Stuff
{
    public static IEnumerable<object?> Items { get; } = new object[2583];
}
```

### Expected behavior

The value of the operand is changed inside of the loop therefore the block is reachable and we expect no warning.

### Actual behavior

Althoug the value of the operand is changed inside of the loop the analyzer assumes the block to be unreachable and we get a warning.

### Known workarounds

Extract the `if`-block to a function.

```csharp
int numberOfNulls = 0;

foreach (object? test in Stuff.Items)
{
    if (test == null)
    {
        numberOfNulls++;

        MaybePlaceholder(numberOfNulls);
    }
}

static void MaybePlaceholder(int numberOfNulls)
{
    if (numberOfNulls % NullsBeforeFeedback == 0)
    {
        Console.WriteLine("placeholder");
    }
}
```

### Related information

* C#/VB.NET Plugins version: SonarAnalyzer.CSharp 9.12.0.78982
* Visual Studio version: 17.8.0
* MSBuild / dotnet version: 17.8.3.51904 / 8.0.100
* SonarScanner for .NET version (if used): -
* Operating System: Windows 11 22H2 (Build 22621.2715)

**Comment by Tim-Pohlmann** [comment]
Hi @E4est thanks for the report. This is a known limitation of the analyzer. We will improve it in the future. The issue is tracked in #8028.
